### ISSUE TITLE
Improve e2e testing

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api-provider-vsphere#2995]
### Improve e2e tests

We compared CAPV govmomi & supervisor e2e tests with core CAPI e2e tests. 
As a result we want to make the following improvements.

Adding e2e tests:
* [x] P0: Add v1.30 => v1.31 upgrade e2e test + check ci-latest (supervisor+govmomi) (@sbueringer)
  * PR: https://github.com/kubernetes/test-infra/pull/32610
  * PR: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2999
  * Plan is to add more and more upgrades jobs for new Kubernetes versions until we have the same test coverage as core CAPI.
* [x] P0: Add test "When testing ClusterClass changes [ClusterClass]" (supervisor) (@zhanggbj)
  * PR: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3011
* [x] P1: Add test "Ensure OwnerReferences and Finalizers are resilient" (supervisor) (@fabriziopandini)
  * [x] CAPI PR https://github.com/kubernetes-sigs/cluster-api/pull/10693
  * [x] CAPV PR https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3059
* [x] P1: Also use `ValidateResourceVersionStable` func (https://github.com/kubernetes-sigs/cluster-api/pull/10530) to ownerRef & finalizer tests (supervisor+govmomi) (@fabriziopandini) https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3033
* [x] P1: Let's check if we have unit test coverage for the following tests for supervisor @chrischdi 
  * capv-e2e.[It] Hardware version upgrade creates a cluster with VM hardware versions upgraded
    * already covered via https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/b935f0aca9e1fa6c8e3d8749f0bfd9a307f9cad7/pkg/services/vmoperator/vmopmachine_test.go#L152
  * capv-e2e.[It] Label nodes with ESXi host info creates a workload cluster whose nodes have the ESXi host info
    * PR: #3085
* [x] P1: Add clusterctl upgrade tests n-2, n-1 (supervisor) (@zhanggbj)
    * CAPV 1.9=>current, CAPI 1.6=>1.7, CAPV 1.10=>current, CAPI 1.7=>1.7 (same we already have on main for govmomi)
    * PR: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3024
* [x] P1: Add test "When testing ClusterClass rollouts [ClusterClass]" (supervisor) (@zhanggbj)
  * PR: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3023
* [x] P1: When upgrading a workload cluster using ClusterClass with RuntimeSDK [ClusterClass] (supervisor) @chrischdi 
    * This requires implementing a minimal Runtime Extension (let's discuss before we start implementing it, and let's do the other more straightforward sub-tasks first)
    * CAPI PR: https://github.com/kubernetes-sigs/cluster-api/pull/10788
    * CAPV PR: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3069
* [x] P1: Audit if we can run more tests with vcsim @chrischdi 
  * #3072
  * #3109
  * #3117

After CAPV v1.11: Adding e2e tests:
* [x] P0: Add clusterctl upgrade tests n-3 (supervisor): https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3159
* [x] P2: Add test "When using the autoscaler with Cluster API using ClusterClass" @chrischdi https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3161

Other improvements:
* [x] Use a kind cluster for clusterctl upgrade tests [CAPI#7613](https://github.com/kubernetes-sigs/cluster-api/issues/7613) (@fabriziopandini)
  * [x] CAPI PR https://github.com/kubernetes-sigs/cluster-api/pull/10639
  * [x] CAPV PR https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3050

Reconsider if we need it later:
* Add test: When using the autoscaler with Cluster API using ClusterClass (incl. scale from zero coverage?)

govmomi has the following additional jobs (intentionally):

<details>

* accepted: it's okay if it's only tested with govmomi:
  * capv-e2e.[It] Cluster creation with [Ignition] bootstrap [PR-Blocking] Should create a workload cluster
* accepted: not surfaced in CAPV supervisor CRDs:
  * capv-e2e.[It] Cluster creation with anti affined nodes should create a cluster with anti-affined nodes
  * capv-e2e.[It] Cluster creation with storage policy should create a cluster successfully
  * capv-e2e.[It] DHCPOverrides configuration test when Creating a cluster with DHCPOverrides configured Only configures the network with the provided nameservers
* accepted: IPAM provider not supported in supervisor mode
  * capv-e2e.[It] ClusterClass Creation using Cluster API quick-start test and IPAM Provider [ClusterClass] Should create a workload cluster
 
</details>

**Comment by zhanggbj** [comment]
I would like to start with 

- [ ] P0: Add test "When testing ClusterClass changes [ClusterClass]" (supervisor)

**Comment by zhanggbj** [comment]
Also would like to pick up a similar one :-)

- [ ] P1: Add test "When testing ClusterClass rollouts [ClusterClass]" (supervisor)

**Comment by zhanggbj** [comment]
It seems this is more about testing efforts. Just raised a PR to test.

- [ ]  P1: Add clusterctl upgrade tests n-2, n-1 (supervisor)

**Comment by sbueringer** [comment]
Yeah good case is that most if it just works once added :)

**Comment by chrischdi** [comment]
Taking this one
- [ ]  P1: Let's check if we have unit test coverage for the following tests for supervisor

**Comment by chrischdi** [comment]
I created a PR for

> - [ ] P1: Let's check if we have unit test coverage for the following tests for supervisor @chrischdi
>   * capv-e2e.[It] Label nodes with ESXi host info creates a workload cluster whose nodes have the ESXi host info
     
https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3085

Regarding the other subpoint:

>   * capv-e2e.[It] Hardware version upgrade creates a cluster with VM hardware versions upgraded

There is already unit-test coverage here:

- https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/b935f0aca9e1fa6c8e3d8749f0bfd9a307f9cad7/pkg/services/vmoperator/vmopmachine_test.go

We define the VSphereMachine here, the called function has a hardcoded value for `vmx-17`:

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/b935f0aca9e1fa6c8e3d8749f0bfd9a307f9cad7/pkg/services/vmoperator/vmopmachine_test.go#L118

Later we check the spec of the VirtualMachine to match the minHardwareVersion (again `17`):

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/b935f0aca9e1fa6c8e3d8749f0bfd9a307f9cad7/pkg/services/vmoperator/vmopmachine_test.go#L152

Plus an immutability check at the end:

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/b935f0aca9e1fa6c8e3d8749f0bfd9a307f9cad7/pkg/services/vmoperator/vmopmachine_test.go#L304-L313

**Comment by sbueringer** [comment]
@chrischdi @fabriziopandini  if I see correctly we don't have the cluster autoscaler test on the list, let's add it as P2?

**Comment by zhanggbj** [comment]
@sbueringer I could take this one once #3024 is merged. Thanks!

- [ ] Add clusterctl upgrade tests n-3 (supervisor)

**Comment by sbueringer** [comment]
> * Add clusterctl upgrade tests n-3

@zhanggbj was just chatting with Christian and Fabrizio about it, we just realized that we didn't have supervisor e2e test in CAPV v1.8. So it would be a lot of effort to add v1.8 => main tests.

We would propose instead to wait until after the v1.11 release in ~ 6 weeks, and then we can just have v1.9 => main, v1.10 => min and v1.11 => main (so we get n-3 coverage just a bit later but with a lot less effort).

Probably we can just include it when we are bumping CAPV main (task "[Prepare main branch for development of the new release](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/docs/release/release-tasks.md#prepare-main-branch-for-development-of-the-new-release)" in https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/3084). Usually we drop the oldest test there, but in that case we will simply keep it around.

**Comment by zhanggbj** [comment]
@sbueringer Makes sense to me. This will provide n-3 tests with much less effort. Thanks!

**Comment by sbueringer** [comment]
@zhanggbj Sorry I was just working on prepare main branch and realized that "Add clusterctl upgrade tests n-3" will be covered by that, so I took over that task from you.

**Comment by zhanggbj** [comment]
@sbueringer Just go ahead, no worries :)

**Comment by sbueringer** [comment]
@chrischdi Assigned `Add test "When using the autoscaler with Cluster API using ClusterClass"` to you + updated the PR description to add "part of ..."

**Comment by sbueringer** [comment]
Moved remaining sub-tasks to separate issues:
* https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/3181
* https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/3182

/close

Great work everyone!
